File: LeetCode/Python Solutions/Task Scheduler/task.py
from typing import List
import heapq


class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        task_frequencies = self.tasksToDic(tasks, {})
        max_heap = []
        for value in task_frequencies.values():
            heapq.heappush(max_heap, -1 * value)
        max_frequency = -1 * heapq.heappop(max_heap)
        max_idle_time = (max_frequency - 1) * n
        while max_heap:
            max_idle_time -= min(max_frequency - 1, -1 * heapq.heappop(max_heap))
        max_idle_time = max(max_idle_time, 0)
        return len(tasks) + max_idle_time

    def tasksToDic(self, tasks, dic):
        for task in tasks:
            if task not in dic:
                dic[task] = 1
            else:
                dic[task] += 1
        return dic


if __name__ == '__main__':
    Instant = Solution()
    Solve = Instant.leastInterval(tasks = ["A","A","A","B","B","B"], n = 2) # tasks = ["A","A","A","B","B","B"], n = 2 -> 8 | tasks = ["A","A","A","B","B","B"], n = 0 -> 6
    print(Solve)

File: LeetCode/Java Solutions/Task Scheduler/task.java
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

class Solution {
    public int leastInterval(char[] tasks, int n) {
        Map<Character, Integer> taskFrequencies = tasksToMap(tasks);
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> b - a);
        maxHeap.addAll(taskFrequencies.values());
        
        int maxFrequency = maxHeap.poll();
        int maxIdleTime = (maxFrequency - 1) * n;
        
        while (!maxHeap.isEmpty()) {
            maxIdleTime -= Math.min(maxFrequency - 1, maxHeap.poll());
        }
        
        maxIdleTime = Math.max(0, maxIdleTime);
        return tasks.length + maxIdleTime;
    }

    private Map<Character, Integer> tasksToMap(char[] tasks) {
        Map<Character, Integer> dic = new HashMap<>();
        for (char task : tasks) {
            dic.put(task, dic.getOrDefault(task, 0) + 1);
        }
        return dic;
    }

    public static void main(String[] args) {
        Solution instant = new Solution();
        int solve = instant.leastInterval(new char[]{'A', 'A', 'A', 'B', 'B', 'B'}, 2);
        System.out.println(solve);  // Expected output: 8
    }
}
